### PR TITLE
Some changes to #6326 (pbkdf2)

### PR DESCRIPTION
**Pull request by Rocknest**

### CODE REVIEW

**Review by andrewrk — APPROVED** [review]
Good to merge when @Rocknest and @rnapier agree on it

**Comment by jedisct1** [comment]
This is great, thanks!

I wouldn't put this in a KDF category, though.

> /// A Key Derivation Function (KDF) is intended to turn a weak, human generated password into a strong key, suitable for cryptographic uses.

A KDF doesn't have to be memory-hard, cache-hard or CPU-hard. HKDF is the most common example of a KDF, and doesn't satisfy any of these properties.

So, maybe it would be better if PBKDF2 would be in a `pwhash` function category instead.

**Comment by Rocknest** [comment]
@jedisct1 what do you think of this doc comment
```zig
/// A Key Derivation Function (KDF) is intended to turn its input (usually
/// weak, human generated password) into a strong key, suitable for cryptographic uses,
/// or a key of a specific length as required by some other cryptographic algorithm.
/// General techniques include salting and stretching the password. Salting injects
/// non-secret random data, so that identical passwords will be converted into unique keys.
/// Stretching applies a deliberately slow hashing function to frustrate brute-force guessing.
/// A KDF usually takes parameters that may affect its cryptographic strength.
```

**Comment by jedisct1** [comment]
Hi @Rocknest !

A KDF derives one or more uniform keys from a master key.

A high-entropy key is usually required, as a KDF is not assumed to provide any mitigations against exhaustive search.

Password stretching functions, on the other hand, intentionally require significant resources to make exhaustive search way more expensive. They can be used as a KDF, as a hash function, or both.

This is why I would recommend leaving the `kdf` namespace for constructions such as HKDF (or even BLAKE3 in KDF mode), and create a `pwhash` (or `pwstretch` or whatever) namespace for constructions explicitly designed for handling passwords, such as PBKDF2.

The comment for that namespace can emphasize the fact that these functions intentionally require significant resources, so that low-entropy keys such as passwords are slow and expensive to break.

**Comment by Rocknest** [comment]
That is perfectly reasonable logic. However since we have stuff like `Md5` in std.crypto i think that the assumption is that the users understand what they're doing.

**Comment by jedisct1** [comment]
`MD5` and `SHA1` shouldn't be used any longer in new applications (except in e.g. the HMAC construction), but they are still required to generate DWARF information or to implement a git client.

Having fine grained categories for cryptographic primitives helps with usability, but also with maintenance. Everything in the same category is expected to have a common (or at least similar) API. A password hashing function doesn't necessarily have the same API as a regular KDF. The number of iterations in PBKDF2 is a parameter that wouldn't make sense in the context of HKDF.

**Comment by jedisct1** [comment]
The implementation is good, thanks a lot for contributing this!

However, the API itself is very specific to PBKDF2 and cannot be reused with other password hashing functions. Having a common API would be useful for consistency and benchmarking purposes.

Instead of directly exposing a `pbkdf2()` function, maybe you could introduce a `PBKDF2` structure parameterized by the MAC function and the number of rounds, and expose a `kdf()` function in that structure.

This will make it easier to later implement other common functions in the same structure, such as handling string-encoded hashes.

File: lib/std/crypto/pbkdf2.zig [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    const c = 1;
+    const dkLen = 1 << 33;
+
+    var derivedKey = try std.testing.allocator.alloc(u8, dkLen);
+    defer {
+        std.testing.allocator.free(derivedKey);
+    }
+
+    try pbkdf2(derivedKey, p, s, c, HmacSha1);
+    // Just verify this doesn't crash with an overflow
```
Comment (jedisct1) [inline lib/std/crypto/pbkdf2.zig:279]: Would you mind verifying that the output eventually matches the expected one?

**Comment by Rocknest** [comment]
@jedisct1 the main author is @rnapier by the way. 
Yeah i agree that the api could be a bit ziggyfied (`Pbkdf2(HmacSha1).deriveKey(...)`).
Since for now we do not have any other KDFs so its okay to have it simply as `crypto.kdf`, later we could move it into `crypto.password_kdf` or other. Generally all crypto API is in development until `1.0`.

File: lib/std/crypto/pbkdf2.zig (hunk 1 above)
Comment (Rocknest) [reply lib/std/crypto/pbkdf2.zig:279]: I cannot run this test due to OOM. Im think ill try to implement mmaped-file allocator to run these kind of tests. Also i dont know what the expected output is. Its few GB long, i think we should try a checksum or something.

**Comment by jedisct1** [comment]
We do have another KDF already (BLAKE3).

Since we know that these have incompatible APIs, it doesn't cost much to put `pbkdf2` in the right category now. 

The whole standard library will not be stable until 1.0, but applications using Zig already exist. We had and will keep having plenty of changes, but the less breaking changes we have, the happier early users will be :)

File: lib/std/crypto/pbkdf2.zig (hunk 1 above)
Comment (jedisct1) [reply lib/std/crypto/pbkdf2.zig:279]: If that test cannot be run, neither on CI nor on developers laptops, maybe it should be removed.

Needing a > 4GB key is completely unrealistic in any real-world scenario. So maybe the assertion in the main function could be simplified to `derivedKey.len < maxInt(u32)`. This doesn't require any testing, and ensures consistent behavior on all platforms.

**Comment by Rocknest** [comment]
Do you have any suggestions for a unified api?

**Comment by jedisct1** [comment]
For password stretching, what is described above would be enough to have a unified API.

To recap, instantiated with PBKDF2 as an example:

- `PBKDF2(Mac, iterations)` returns a structure
- That structure includes a `kdf()` (or `deriveKey()`) function

The only important part is to keep PBKDF2-specific parameters out of the `kdf()` function. PBKDF2 only requires a number of iterations, but other functions require a different parameter set (memory, parallelism, key, etc).

For storage, the output of password hashing functions along with the parameters required for verification are usually encoded as a string, and these encodings are more or less standardized. This can be added later, as long as each password hashing function is represented by a type.

**Comment by rnapier** [comment]
Been away, lots of comments :D

> So, maybe it would be better if PBKDF2 would be in a `pwhash` function category instead.

Yeah. That's reasonable. I originally called the namespace `stretch` (but I think `pwhash` is clearer). I agree that HKDF and PBKDF2 are very different creatures, and HKDF is more clearly a "KDF".

>That is perfectly reasonable logic. However since we have stuff like Md5 in std.crypto i think that the assumption is that the users understand what they're doing.

I don't think we should assume that developers know what they're doing with crypto primitives. They don't. I spend a lot of time fixing people's crypto mistakes in a lot of different languages because so many languages make it easy to do it wrong.

Crypto APIs should be easy to use correctly and more difficult to use incorrectly. A classic failure example isa default value of all-0s for the IV in block cipher modes. That's bad API design.

That does not mean that "unsafe" primitives should be removed. They're important to be available both for compatibility and because there are many uses where a so-called "unsafe" primitive is completely fine. I think there's some value to marking deeply broken primitives like MD5 and easy to misuse primitives like ECB with some kind of "insecure" marking. Though documentation may be sufficient rather than a separate namespace.

> Instead of directly exposing a pbkdf2() function, maybe you could introduce a PBKDF2 structure parameterized by the MAC function and the number of rounds, and expose a kdf() function in that structure.

Reasonable suggestion. I wasn't sure what the normal Zig approach here was. I like Rocknest's `deriveKey()` name better than `kdf()` for this.

I think the best way to work out the API is to implement a second pwhash and see where the two share and differ. I'm thinking bcrypt.

@andrewrk Do you prefer to merge this and have a new PR to improve it, or wait for a better version? I'm happy to pull these recommendations together and submit a new patch either way, unless @Rocknest would rather do it (they're obviously more experienced with Zig than I am).

**Comment by andrewrk** [comment]
I'm happy to merge this now, I just am kind of unqualified to review it so I wanted to make sure the crypto folks around here had a chance to have a look and make sure there are no footguns :+1:

**Comment by Rocknest** [comment]
I've extracted some suggestions into #6371

**Comment by rnapier** [comment]
Excellent. I'll start looking at it this weekend. I'm glad to see a good collection of crypto-aware folks to keep me honest. Security is a collaborative exercise.

**Pull request merged**